examples/getitem/src/lib.rs

```diff
@@ -0,0 +1,55 @@
+// This is a very fake example of how to check __getitem__ parameter and handle appropriately
+use pyo3::prelude::*;
+use pyo3::exceptions::PyTypeError;
+
+
+#[pyclass]
+struct ExampleContainer {
+
+}
+
+#[pymethods]
+impl ExampleContainer {
+    #[new]
+    fn new() -> Self {
+        ExampleContainer{}
+    }
+
+    fn __getitem__(&self, key: Py<PyAny>, _py: Python) -> PyResult<i32>{
+        // test
+        let key_ref: &PyAny = &key.as_ref(_py);
+        let key_type = key_ref.get_type();
+        let type_name = key_type.name().unwrap();
+        let result = match type_name {
```

Comment:
I think this should not match on the type name, rather trying to extract what is expected and retry otherwise, e.g.

```rust
if let Ok(position) = key.extract::<i32>() {
  ..
} else if let Ok(slice) = key.extract::<PySlice>() {
  ..
} else {
  Err(PyTypeError::new_err("Unsupported type"))
}
```

Reply:
I never could get extract to work for PySlice:

```
27  |         } else if let Ok(slice) = key.extract::<PySlice>(_py){
    |                                                 ^^^^^^^ the trait `PyClass` is not implemented for `PySlice`
```

Reply:
You probably need to replace `extract` by `downcast` then.

Reply:
> I still needed getattr. It wasn't intuitive to me how to get from PySlice to PySliceIndices (https://github.com/PyO3/pyo3/blob/main/src/types/slice.rs)...or if that was really what was the right thing....hopefully gettattr is still the right thing here.

I think you want to call [`PySlice::indices`](https://docs.rs/pyo3/latest/pyo3/types/struct.PySlice.html#method.indices) by passing in the conceptual length of your container. (You should probably amend your example to actually store a container like `Vec<i32>` created via `(0..length).collect()`.) The methods needs to know the length to be able to handle negative indices "counting from the end."

> Is it right that:

Yes, that is basically correct.

Reply:
Personally, I am not a fan of showing both usage of `indices` and manually calls to `getattr` as it might confuse users even more. But let's see what other people think...

Reply:
It seems a lot of the examples currently showed getattr in the docs so that's how I started, but I agree that indices is better for 95% of the use cases. I could just see someone needing the original knowledge of "no end given" or "no step given" just to make sure we have same functionality here that is available in python.

If it is confusing, I could do two separate containers with each as an example?

Add links to the docs we've referenced so far as well....